# Домашнее задание Лист 03

## 1.1.1

In [2]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [3]:
data_string='''sleep totwrk age male hrwage
3113   3438  32   1    7.07 
2920   5020  31   1    1.43 
2670   2815  44   1   20.53 
3083   3786  30   0    9.62 
3448   2580  64   1    2.75 
4063   1205  41   1   19.25'''

In [4]:
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

sleep  totwrk  age  male  hrwage
0   3113    3438   32     1    7.07
1   2920    5020   31     1    1.43
2   2670    2815   44     1   20.53
3   3083    3786   30     0    9.62
4   3448    2580   64     1    2.75
5   4063    1205   41     1   19.25

In [5]:
df['y_pred'] =  3525.14 + df['totwrk']*(-0.16) + df['age']*(1.62) + df['male']*(51.84) + np.log(df['hrwage'])*(-9.88)
df

sleep  totwrk  age  male  hrwage       y_pred
0   3113    3438   32     1    7.07  3059.416098
1   2920    5020   31     1    1.43  2820.466176
2   2670    2815   44     1   20.53  3168.003754
3   3083    3786   30     0    9.62  2945.613219
4   3448    2580   64     1    2.75  3257.865383
5   4063    1205   41     1   19.25  3421.379791

In [6]:
df['residual']=df['sleep']-df['y_pred']
df

sleep  totwrk  age  male  hrwage       y_pred    residual
0   3113    3438   32     1    7.07  3059.416098   53.583902
1   2920    5020   31     1    1.43  2820.466176   99.533824
2   2670    2815   44     1   20.53  3168.003754 -498.003754
3   3083    3786   30     0    9.62  2945.613219  137.386781
4   3448    2580   64     1    2.75  3257.865383  190.134617
5   4063    1205   41     1   19.25  3421.379791  641.620209

## 1.1.2

In [7]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [8]:
data_string2='''output capital labour
9.25    2.61    184  
3.66    1.32     91  
28.78   22.09   426  
4.12    10.74    72  
2.89    1.16     46  '''
df = pd.read_csv(io.StringIO(data_string2), sep='\s+')
df

output  capital  labour
0    9.25     2.61     184
1    3.66     1.32      91
2   28.78    22.09     426
3    4.12    10.74      72
4    2.89     1.16      46

In [9]:
df['y_pred'] =  -1.71  + np.log(df['capital'])*(0.21) + np.log(df['labour'])*(0.71)
df

output  capital  labour    y_pred
0    9.25     2.61     184  2.194068
1    3.66     1.32      91  1.551013
2   28.78    22.09     426  3.238628
3    4.12    10.74      72  1.824968
4    2.89     1.16      46  1.039504

In [10]:
df['residual']=np.log(df['output'])-df['y_pred']
df. round (2)

output  capital  labour  y_pred  residual
0    9.25     2.61     184    2.19      0.03
1    3.66     1.32      91    1.55     -0.25
2   28.78    22.09     426    3.24      0.12
3    4.12    10.74      72    1.82     -0.41
4    2.89     1.16      46    1.04      0.02

## 1.2.1

In [11]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [12]:
TSS=100.35
ESS=47.56

In [13]:
R2=ESS/TSS
round(R2,2)

0.47

In [14]:
RSS=TSS-ESS
round (RSS,2)

52.79

###
$$
R^2 = 52.79, RSS = 0,47 
$$


## 1.2.2

In [15]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [16]:
TSS=240.82
RSS=93.52
ESS=TSS-RSS
ESS

147.3

In [17]:
R2=ESS/TSS
round(R2,2)

0.61

###
$$
ESS = 147.3 , R^2 = 0.61
$$

## 1.2.3

In [20]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [21]:
ESS=418.38
RSS=782.83
TSS=ESS+RSS
round(TSS,2)

1201.21

In [22]:
R2=ESS/TSS
round(R2,2)

0.35

###
$$
TSS = 1201.21 , R^2 = 0.35
$$

## 1.2.4 

In [24]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [25]:
data_string='''residual
-0.24
0.41
1.22
-0.93
-0.37
-1.33
0.16
-0.27
0.06
1.29
'''
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df['residual']
df

residual
0     -0.24
1      0.41
2      1.22
3     -0.93
4     -0.37
5     -1.33
6      0.16
7     -0.27
8      0.06
9      1.29

In [26]:
# создаем столбец со значениями квадратов регресионных остатков
df['residual^2'] = df['residual']**2
df

residual  residual^2
0     -0.24      0.0576
1      0.41      0.1681
2      1.22      1.4884
3     -0.93      0.8649
4     -0.37      0.1369
5     -1.33      1.7689
6      0.16      0.0256
7     -0.27      0.0729
8      0.06      0.0036
9      1.29      1.6641

In [27]:
# находим сумму квадратов регрессионных остатков
df[['residual^2']].sum(axis=0)

residual^2    6.251
dtype: float64

###
$$
RSS = 6.25
$$

### Ошибка регрессии вычисляется по формуле:

###
$$
s = \sqrt {(\frac {RSS}{n-k-1})}
$$

In [29]:
n=10
k=2
s=np.sqrt(6.25/(n-k-1))
round(s,2)

0.94

###
$$
s = 0.94
$$

## 1.2.5

In [30]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [31]:
data_string='''sleep/60 totwrk age y.hat
51.88     3438  32  51.00
48.67     5020  31  44.43
44.50     2815  44  53.72
51.38     3786  30  49.53
57.47     2580  64  54.92
67.72     1205  41  60.36
53        2113  35  56.53
48.80     3608  47  50.46
56.13     2353  32  55.50
50.30     2851  30  53.41
'''
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

sleep/60  totwrk  age  y.hat
0     51.88    3438   32  51.00
1     48.67    5020   31  44.43
2     44.50    2815   44  53.72
3     51.38    3786   30  49.53
4     57.47    2580   64  54.92
5     67.72    1205   41  60.36
6     53.00    2113   35  56.53
7     48.80    3608   47  50.46
8     56.13    2353   32  55.50
9     50.30    2851   30  53.41

In [32]:
# Общая сумма квадратов TSS:
TSS = ((df['sleep/60'] - np.mean(df['sleep/60']))**2).sum()
TSS.round(2)

366.26

In [33]:
# Остаточная сумма квадртов
RSS = ((df['sleep/60'] - df['y.hat'])**2).sum()
RSS.round(2)

193.14

In [34]:
# Объясненная сумма квадратов, в математическом смысле НЕ совпадает с ответом
ESS=TSS-RSS
round(ESS,2)

173.12

In [35]:
# Объясненная сумма квадратов по ФОРМУЛЕ
ESS = ((df['y.hat'] - np.mean(df['sleep/60']))**2).sum()
ESS.round(2)

173.19

###
$$
TSS = 366.26, RSS = 193.14, ESS = 173.19 
$$

## 1.2.6

In [36]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [37]:
data_string='''log(output) log(capital) log(labour) y.hat
2.22            0.96        5.21     2.19 
1.30            0.28        4.51     1.47 
3.36            3.10        6.05     3.23 
1.42            2.37        4.28     1.56 
1.06            0.15        3.83     0.85 
6.17            5.62        8.94     6.15 
0.77            0.52        3.87     0.94 
1.42            0.15        4.68     1.61 
0.90           -0.13        3.56     0.57 
1.35            1.25        4.28     1.40 
'''
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

log(output)  log(capital)  log(labour)  y.hat
0         2.22          0.96         5.21   2.19
1         1.30          0.28         4.51   1.47
2         3.36          3.10         6.05   3.23
3         1.42          2.37         4.28   1.56
4         1.06          0.15         3.83   0.85
5         6.17          5.62         8.94   6.15
6         0.77          0.52         3.87   0.94
7         1.42          0.15         4.68   1.61
8         0.90         -0.13         3.56   0.57
9         1.35          1.25         4.28   1.40

In [38]:
# Общая сумма квадратов TSS:
TSS = ((df['log(output)'] - np.mean(df['log(output)']))**2).sum()
TSS.round(2)

24.48

In [39]:
# Остаточная сумма квадртов
RSS = ((df['log(output)'] - df['y.hat'])**2).sum()
RSS.round(2)

0.29

In [40]:
# Объясненная сумма квадратов по ФОРМУЛЕ
ESS = ((df['y.hat'] - np.mean(df['log(output)']))**2).sum()
ESS.round(2)

24.25

###
$$
TSS = 24.48, RSS = 0.29, ESS = 24.25
$$

## 1.2.7

In [41]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [42]:
n=935
k=5
s=0.378

###
$$
s = \sqrt {(\frac {RSS}{n-k-1})}
$$

## Выше формула вычесления ошибки регрессии, значит RSS находим по формуле:

###
$$
RSS = (n-k-1) * s^2
$$

In [44]:
RSS=(n-k-1)*(s**2)
round(RSS,3)

132.739

###
$$
RSS = 132.739
$$

## 1.2.8

In [45]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [46]:
n=935
k=4
s=0.383
R2=0.178

##
$$
RSS^2 = 1 - (\frac {RSS}{TSS})
$$

## 
$$
TSS = \frac {RSS}{1-R^2} = \frac {(n-k-1)*s^2}{1-R^2}
$$

In [48]:
TSS=((n-k-1)*s**2)/(1-R2)
round(TSS,3)

165.962

##
$$
TSS = 165.962
$$

## 1.2.9

In [49]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [50]:
n=935
k=3
R2=0.149
s=0.389

##
$$
RSS = (n - k - 1) * s^2
$$

In [52]:
RSS=(n-k-1)*(s**2)
round(RSS,3)

140.88

##
$$
RSS = 140.88
$$

##
$$
RSS^2 = 1 - (\frac {RSS}{TSS})
$$

$$
TSS = \frac {RSS}{1-R^2} = \frac {(n-k-1)*s^2}{1-R^2}
$$

In [53]:
TSS=((n-k-1)*s**2)/(1-R2)
round(TSS,3)

165.546

##
$$
ESS = TSS - RSS
$$

In [54]:
ESS=TSS-RSS
round(ESS,3)

24.666

##
$$
ESS = 24.666
$$

## 2.1

In [55]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

##
$$
R^2_{adj} = 1 - (1 - R^2) * \frac{n -1}{n - k - 1}
$$

In [56]:
n=526 
k=5 
R2=0.288

In [57]:
R2adj=1-(1-R2)*(n-1)/(n-k-1)
round(R2adj,3)

0.281

##
$$
R^2_{adj} = 0.281
$$

## 2.2

In [59]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

##
$$
R^2_{adj} = 1 - (1 - R^2) * \frac{n -1}{n - k - 1}
$$

In [60]:
n=935 
k=6
R2=0.201

In [61]:
R2adj=1-(1-R2)*(n-1)/(n-k-1)
round(R2adj,3)

0.196

##
$$
R^2_{adj} = 0.196
$$

## 2.3

In [62]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [63]:
data_string='''sleep/60 totwrk age y.hat
51.88     3438  32  51.00
48.67     5020  31  44.43
44.50     2815  44  53.72
51.38     3786  30  49.53
57.47     2580  64  54.92
67.72     1205  41  60.36
53        2113  35  56.53
48.80     3608  47  50.46
56.13     2353  32  55.50
50.30     2851  30  53.41
'''
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

sleep/60  totwrk  age  y.hat
0     51.88    3438   32  51.00
1     48.67    5020   31  44.43
2     44.50    2815   44  53.72
3     51.38    3786   30  49.53
4     57.47    2580   64  54.92
5     67.72    1205   41  60.36
6     53.00    2113   35  56.53
7     48.80    3608   47  50.46
8     56.13    2353   32  55.50
9     50.30    2851   30  53.41

In [64]:
ESS = ((df['y.hat'] - np.mean(df['sleep/60']))**2).sum()
ESS.round(2)

173.19

In [65]:
TSS = ((df['sleep/60'] - np.mean(df['sleep/60']))**2).sum()
TSS.round(2)

366.26

In [66]:
R2=ESS/TSS
round(R2,2)

0.47

In [67]:
n=10 
k=2

In [68]:
R2adj=1-(1-R2)*(n-1)/(n-k-1)
round(R2adj,2)

0.32

##
$$
R^2_{adj} = 0.32
$$

## 2.4

In [70]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [71]:
data_string='''log(output) log(capital) log(labour) y.hat
2.22            0.96        5.21     2.19 
1.30            0.28        4.51     1.47 
3.36            3.10        6.05     3.23 
1.42            2.37        4.28     1.56 
1.06            0.15        3.83     0.85 
6.17            5.62        8.94     6.15 
0.77            0.52        3.87     0.94 
1.42            0.15        4.68     1.61 
0.90           -0.13        3.56     0.57 
1.35            1.25        4.28     1.40
'''
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

log(output)  log(capital)  log(labour)  y.hat
0         2.22          0.96         5.21   2.19
1         1.30          0.28         4.51   1.47
2         3.36          3.10         6.05   3.23
3         1.42          2.37         4.28   1.56
4         1.06          0.15         3.83   0.85
5         6.17          5.62         8.94   6.15
6         0.77          0.52         3.87   0.94
7         1.42          0.15         4.68   1.61
8         0.90         -0.13         3.56   0.57
9         1.35          1.25         4.28   1.40

In [72]:
ESS = ((df['y.hat'] - np.mean(df['log(output)']))**2).sum()
ESS.round(2)

24.25

In [73]:
TSS = ((df['log(output)'] - np.mean(df['log(output)']))**2).sum()
TSS.round(2)

24.48

In [74]:
R2=ESS/TSS
round(R2,2)

0.99

In [75]:
n=10
k=2
R2adj=1-(1-R2)*(n-1)/(n-k-1)
round(R2adj,2)

0.99

##
$$
R^2_{adj} = 0.99
$$

## 3.1

In [78]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [79]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     ...  spwrk75  totwrk  union  worknrm  workscnd  exper  yngkid  yrsmarr  \
0    ...        0    3438      0     3438         0     14       0       13   
1    ...        0    5020      0     5020         0     11       0        0   
2    ...        1    2815      0     2815         0     21       0        0   
3    ...        1    3786      0     3786         0     12       0       12   
4    ...        1    2580      0     2580         0     44       0       33   
..   ...      ...     ...    ...      ...       ...    ...     ...      ...   
701  ...        1    2026      0     2026         0     27       0       18   
702  ...        0     675      1      465       210     18       0        4   
703  ...        1    1851      0     1851         0     19       0       17   
704  ...        1    1961      1     1481       480     31       0       22   
705  ...        0    2363      0     2363         0      8       1        9   

        hrwage  agesq  
0     7.070004   1024  
1     1.429999    961  
2    20.529997   1936  
3     9.619998    900  
4     2.750000   4096  
..         ...    ...  
701        NaN   2025  
702        NaN   1156  
703        NaN   1369  
704        NaN   2916  
705        NaN    900  

[706 rows x 34 columns]

In [80]:
sleep_eq1 = smf.ols(formula='sleep~totwrk+age+male+smsa', data=df).fit()
sleep_eq1.params

Intercept    3494.223143
totwrk         -0.167681
age             2.806498
male           86.908377
smsa          -75.285755
dtype: float64

##
$$
Intercept = 3494.223, totwrk = 0.17, age = 2.81 , male = 86.91 , smsa = -75.29
$$

In [81]:
sleep_eq1.ess.round(3)
# ESS

17188382.234

In [82]:
sleep_eq1.centered_tss.round(3)
# TSS

139239835.763

In [83]:
RSS=sleep_eq1.ssr.round(3)
RSS
# RSS

122051453.53

In [84]:
sleep_eq1.rsquared.round(7)
# R2

0.1234444

In [85]:
sleep_eq1.rsquared_adj.round(7)
# R2adj

0.1184427

In [86]:
n=706
k=4
s=np.sqrt(RSS/(n-k-1))
round(s,4)

417.2655

## 3.2

In [87]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [88]:
pd.set_option('display.max_columns', None)

In [89]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     leis2  leis3  smsa   lhrwage    lothinc  male  marr  prot  rlxall  selfe  \
0     3479   3479     0  1.955861  10.075380     1     1     1    3163      0   
1     2140   2140     0  0.357674   0.000000     1     0     1    2920      1   
2     4505   4227     1  3.021887   0.000000     1     1     0    3038      1   
3     3211   3211     0  2.263844   0.000000     0     1     1    3083      1   
4     4007   4007     0  1.011601   9.328213     1     1     1    3493      0   
..     ...    ...   ...       ...        ...   ...   ...   ...     ...    ...   
701   4669   4644     1       NaN   9.664660     0     1     1    3410      0   
702   5870   5870     0       NaN   0.000000     1     1     0    3535      0   
703   4719   4584     0       NaN   9.235130     0     1     1    3645      0   
704   5119   5031     0       NaN   9.210441     0     1     1    3088      1   
705   4302   4302     0       NaN   9.940350     0     1     0    3415      0   

     sleep  slpnaps  south  spsepay  spwrk75  totwrk  union  worknrm  \
0     3113     3163      0        0        0    3438      0     3438   
1     2920     2920      1        0        0    5020      0     5020   
2     2670     2760      0    20000        1    2815      0     2815   
3     3083     3083      0     5000        1    3786      0     3786   
4     3448     3493      0     2400        1    2580      0     2580   
..     ...      ...    ...      ...      ...     ...    ...      ...   
701   2985     3385      0    16000        1    2026      0     2026   
702   3520     3535      0        0        0     675      1      465   
703   3510     3510      0    12000        1    1851      0     1851   
704   2970     3000      1    35000        1    1961      1     1481   
705   2970     3415      1        0        0    2363      0     2363   

     workscnd  exper  yngkid  yrsmarr     hrwage  agesq  
0           0     14       0       13   7.070004   1024  
1           0     11       0        0   1.429999    961  
2           0     21       0        0  20.529997   1936  
3           0     12       0       12   9.619998    900  
4           0     44       0       33   2.750000   4096  
..        ...    ...     ...      ...        ...    ...  
701         0     27       0       18        NaN   2025  
702       210     18       0        4        NaN   1156  
703         0     19       0       17        NaN   1369  
704       480     31       0       22        NaN   2916  
705         0      8       1        9        NaN    900  

[706 rows x 34 columns]

In [92]:
df = df[df.smsa >= 1]
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
7     47      0     8  0.000000  0.000000    13        0       1     1   3544   
9     30      0    10  0.000000  0.000000    15    11750       1     1   4211   
11    23      0    12  0.000000  0.000000    16     1000       1     1   6415   
12    24      0    13  0.000000  0.000000    16     1000       1     1   3844   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
689   31      0   690  0.182331  0.030075    12    10500       1     0   4639   
693   25      1   694  0.182331  0.030075    12     3500       1     0   5530   
697   47      0   698  0.182331  0.030075    10     9500       1     0   5882   
700   32      0   701  0.182331  0.030075    16    10500       1     0   4707   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   

     leis2  leis3  smsa   lhrwage    lothinc  male  marr  prot  rlxall  selfe  \
2     4505   4227     1  3.021887   0.000000     1     1     0    3038      1   
7     3469   3439     1  2.708717  10.657280     1     1     1    3033      1   
9     4061   4061     1  1.085189   8.853808     1     1     0    3168      1   
11    6355   6230     1  0.207014   7.313887     0     1     0    3480      0   
12    3844   3794     1  0.207014   7.313887     1     1     0    3848      0   
..     ...    ...   ...       ...        ...   ...   ...   ...     ...    ...   
689   4639   4489     1       NaN   9.282754     0     0     0    3218      0   
693   4845   4845     1       NaN   0.000000     0     0     1    3725      0   
697   5747   5747     1       NaN   9.564583     0     1     0    3150      0   
700   4587   4512     1       NaN   9.282754     0     0     0    3180      0   
701   4669   4644     1       NaN   9.664660     0     1     1    3410      0   

     sleep  slpnaps  south  spsepay  spwrk75  totwrk  union  worknrm  \
2     2670     2760      0    20000        1    2815      0     2815   
7     2928     3003      0        0        0    3608      0     3608   
9     3018     3168      0     6000        1    2851      0     2851   
11    3295     3355      0     3000        1     370      0      370   
12    3798     3798      0     3000        1    2438      0     2438   
..     ...      ...    ...      ...      ...     ...    ...      ...   
689   3068     3068      1        0        0    2373      0     2373   
693   3040     3725      0        0        0    1510      1     1510   
697   3015     3150      0        0        0    1183      1     1183   
700   2985     3105      1        0        0    2388      0     2388   
701   2985     3385      0    16000        1    2026      0     2026   

     workscnd  exper  yngkid  yrsmarr     hrwage  agesq  
2           0     21       0        0  20.529997   1936  
7           0     28       0       24  15.010007   2209  
9           0      9       0        7   2.959999    900  
11          0      1       0        4   1.230000    529  
12          0      2       0        4   1.230000    576  
..        ...    ...     ...      ...        ...    ...  
689         0     13       0        0        NaN    961  
693         0      7       1        0        NaN    625  
697         0     31       0       28        NaN   2209  
700         0     10       0        0        NaN   1024  
701         0     27       0       18        NaN   2025  

[282 rows x 34 columns]

In [93]:
sleep_eq1 = smf.ols(formula='sleep~totwrk+age+male+south', data=df).fit()
sleep_eq1.params

Intercept    3460.427182
totwrk         -0.184996
age             2.276770
male           92.649951
south         168.852598
dtype: float64

In [94]:
sleep_eq1.ess.round(3)
# ESS

7396259.393

In [95]:
sleep_eq1.centered_tss.round(3)
# TSS

50055043.404

In [98]:
RSS=sleep_eq1.ssr.round(3)
RSS
# RSS

42658784.011

In [99]:
sleep_eq1.rsquared.round(7)
# R2

0.1477625

In [100]:
sleep_eq1.rsquared_adj.round(7)
# R2adj

0.1354558

In [101]:
n=282
k=4
s=np.sqrt(RSS/(n-k-1))
round(s,4)

392.4319

## 3.3

In [102]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [104]:
pd.set_option('display.max_columns', None)

In [105]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     leis2  leis3  smsa   lhrwage    lothinc  male  marr  prot  rlxall  selfe  \
0     3479   3479     0  1.955861  10.075380     1     1     1    3163      0   
1     2140   2140     0  0.357674   0.000000     1     0     1    2920      1   
2     4505   4227     1  3.021887   0.000000     1     1     0    3038      1   
3     3211   3211     0  2.263844   0.000000     0     1     1    3083      1   
4     4007   4007     0  1.011601   9.328213     1     1     1    3493      0   
..     ...    ...   ...       ...        ...   ...   ...   ...     ...    ...   
701   4669   4644     1       NaN   9.664660     0     1     1    3410      0   
702   5870   5870     0       NaN   0.000000     1     1     0    3535      0   
703   4719   4584     0       NaN   9.235130     0     1     1    3645      0   
704   5119   5031     0       NaN   9.210441     0     1     1    3088      1   
705   4302   4302     0       NaN   9.940350     0     1     0    3415      0   

     sleep  slpnaps  south  spsepay  spwrk75  totwrk  union  worknrm  \
0     3113     3163      0        0        0    3438      0     3438   
1     2920     2920      1        0        0    5020      0     5020   
2     2670     2760      0    20000        1    2815      0     2815   
3     3083     3083      0     5000        1    3786      0     3786   
4     3448     3493      0     2400        1    2580      0     2580   
..     ...      ...    ...      ...      ...     ...    ...      ...   
701   2985     3385      0    16000        1    2026      0     2026   
702   3520     3535      0        0        0     675      1      465   
703   3510     3510      0    12000        1    1851      0     1851   
704   2970     3000      1    35000        1    1961      1     1481   
705   2970     3415      1        0        0    2363      0     2363   

     workscnd  exper  yngkid  yrsmarr     hrwage  agesq  
0           0     14       0       13   7.070004   1024  
1           0     11       0        0   1.429999    961  
2           0     21       0        0  20.529997   1936  
3           0     12       0       12   9.619998    900  
4           0     44       0       33   2.750000   4096  
..        ...    ...     ...      ...        ...    ...  
701         0     27       0       18        NaN   2025  
702       210     18       0        4        NaN   1156  
703         0     19       0       17        NaN   1369  
704       480     31       0       22        NaN   2916  
705         0      8       1        9        NaN    900  

[706 rows x 34 columns]

In [106]:
df = df[df.male >= 1]
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
5     41      0     6  0.000000  0.000000    12        0       1     1   4812   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
692   50      0   693  0.182331  0.030075    12     2500       1     0   4417   
694   44      0   695  0.182331  0.030075    16    16250       1     0   4539   
695   29      0   696  0.182331  0.030075    16    16250       1     0   4639   
699   61      0   700  0.182331  0.030075    17    27500       1     0   4362   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   

     leis2  leis3  smsa   lhrwage    lothinc  male  marr  prot  rlxall  selfe  \
0     3479   3479     0  1.955861  10.075380     1     1     1    3163      0   
1     2140   2140     0  0.357674   0.000000     1     0     1    2920      1   
2     4505   4227     1  3.021887   0.000000     1     1     0    3038      1   
4     4007   4007     0  1.011601   9.328213     1     1     1    3493      0   
5     4797   4797     0  2.957511  10.657280     1     1     1    4078      0   
..     ...    ...   ...       ...        ...   ...   ...   ...     ...    ...   
692   4147   4147     0       NaN   9.328213     1     1     0    3330      0   
694   4451   4393     0       NaN   7.824446     1     1     1    3174      0   
695   4519   4519     0       NaN   9.328213     1     1     1    3143      0   
699   4362   4237     0       NaN   9.615872     1     1     0    3118      1   
702   5870   5870     0       NaN   0.000000     1     1     0    3535      0   

     sleep  slpnaps  south  spsepay  spwrk75  totwrk  union  worknrm  \
0     3113     3163      0        0        0    3438      0     3438   
1     2920     2920      1        0        0    5020      0     5020   
2     2670     2760      0    20000        1    2815      0     2815   
4     3448     3493      0     2400        1    2580      0     2580   
5     4063     4078      0        0        0    1205      0        0   
..     ...      ...    ...      ...      ...     ...    ...      ...   
692   3060     3330      0     2500        1    2603      0     2603   
694   3028     3116      0        0        0    2513      0     2513   
695   3023     3143      0        0        0    2418      0     2418   
699   2993     2993      0        0        0    2725      0     2725   
702   3520     3535      0        0        0     675      1      465   

     workscnd  exper  yngkid  yrsmarr     hrwage  agesq  
0           0     14       0       13   7.070004   1024  
1           0     11       0        0   1.429999    961  
2           0     21       0        0  20.529997   1936  
4           0     44       0       33   2.750000   4096  
5        1205     23       0       23  19.249998   1681  
..        ...    ...     ...      ...        ...    ...  
692         0     32       0       22        NaN   2500  
694         0     22       0       19        NaN   1936  
695         0      7       1        5        NaN    841  
699         0     38       0       30        NaN   3721  
702       210     18       0        4        NaN   1156  

[400 rows x 34 columns]

In [107]:
sleep_eq1 = smf.ols(formula='sleep~totwrk+age+smsa+south', data=df).fit()
sleep_eq1.params

Intercept    3600.958512
totwrk         -0.187650
age             3.301936
smsa          -77.686142
south          60.113665
dtype: float64

In [108]:
sleep_eq1.ess.round(3)
# ESS

12125135.598

In [109]:
sleep_eq1.centered_tss.round(3)
# TSS

75570140.578

In [110]:
RSS=sleep_eq1.ssr.round(3)
RSS
# RSS

63445004.98

In [111]:
sleep_eq1.rsquared.round(7)
# R2

0.1604488

In [112]:
sleep_eq1.rsquared_adj.round(7)
# R2adj

0.151947

In [113]:
n=400
k=4
s=np.sqrt(RSS/(n-k-1))
round(s,4)

400.7746

## 3.4

In [114]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [115]:
pd.set_option('display.max_columns', None)

In [116]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     leis2  leis3  smsa   lhrwage    lothinc  male  marr  prot  rlxall  selfe  \
0     3479   3479     0  1.955861  10.075380     1     1     1    3163      0   
1     2140   2140     0  0.357674   0.000000     1     0     1    2920      1   
2     4505   4227     1  3.021887   0.000000     1     1     0    3038      1   
3     3211   3211     0  2.263844   0.000000     0     1     1    3083      1   
4     4007   4007     0  1.011601   9.328213     1     1     1    3493      0   
..     ...    ...   ...       ...        ...   ...   ...   ...     ...    ...   
701   4669   4644     1       NaN   9.664660     0     1     1    3410      0   
702   5870   5870     0       NaN   0.000000     1     1     0    3535      0   
703   4719   4584     0       NaN   9.235130     0     1     1    3645      0   
704   5119   5031     0       NaN   9.210441     0     1     1    3088      1   
705   4302   4302     0       NaN   9.940350     0     1     0    3415      0   

     sleep  slpnaps  south  spsepay  spwrk75  totwrk  union  worknrm  \
0     3113     3163      0        0        0    3438      0     3438   
1     2920     2920      1        0        0    5020      0     5020   
2     2670     2760      0    20000        1    2815      0     2815   
3     3083     3083      0     5000        1    3786      0     3786   
4     3448     3493      0     2400        1    2580      0     2580   
..     ...      ...    ...      ...      ...     ...    ...      ...   
701   2985     3385      0    16000        1    2026      0     2026   
702   3520     3535      0        0        0     675      1      465   
703   3510     3510      0    12000        1    1851      0     1851   
704   2970     3000      1    35000        1    1961      1     1481   
705   2970     3415      1        0        0    2363      0     2363   

     workscnd  exper  yngkid  yrsmarr     hrwage  agesq  
0           0     14       0       13   7.070004   1024  
1           0     11       0        0   1.429999    961  
2           0     21       0        0  20.529997   1936  
3           0     12       0       12   9.619998    900  
4           0     44       0       33   2.750000   4096  
..        ...    ...     ...      ...        ...    ...  
701         0     27       0       18        NaN   2025  
702       210     18       0        4        NaN   1156  
703         0     19       0       17        NaN   1369  
704       480     31       0       22        NaN   2916  
705         0      8       1        9        NaN    900  

[706 rows x 34 columns]

In [117]:
df = df[df.male <= 0]
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
11    23      0    12  0.000000  0.000000    16     1000       1     1   6415   
27    35      0    28  1.000000  0.000000    12     6750       1     1   4904   
42    44      0    43  0.000000  0.000000    17     6750       1     1   4804   
55    24      0    56  0.000000  0.000000    14     1000       1     1   5490   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
700   32      0   701  0.182331  0.030075    16    10500       1     0   4707   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     leis2  leis3  smsa   lhrwage   lothinc  male  marr  prot  rlxall  selfe  \
3     3211   3211     0  2.263844  0.000000     0     1     1    3083      1   
11    6355   6230     1  0.207014  7.313887     0     1     0    3480      0   
27    4829   4829     1  0.989541  7.313887     0     0     1    3001      0   
42    4591   4591     0  0.770108  7.313887     0     0     0    2978      1   
55    5490   5490     0  0.048790  0.000000     0     0     0    2790      1   
..     ...    ...   ...       ...       ...   ...   ...   ...     ...    ...   
700   4587   4512     1       NaN  9.282754     0     0     0    3180      0   
701   4669   4644     1       NaN  9.664660     0     1     1    3410      0   
703   4719   4584     0       NaN  9.235130     0     1     1    3645      0   
704   5119   5031     0       NaN  9.210441     0     1     1    3088      1   
705   4302   4302     0       NaN  9.940350     0     1     0    3415      0   

     sleep  slpnaps  south  spsepay  spwrk75  totwrk  union  worknrm  \
3     3083     3083      0     5000        1    3786      0     3786   
11    3295     3355      0     3000        1     370      0      370   
27    2926     3001      0        0        0    2250      0     2250   
42    2765     2978      0        0        0    2511      0     2511   
55    2790     2790      0        0        0    1800      1     1800   
..     ...      ...    ...      ...      ...     ...    ...      ...   
700   2985     3105      1        0        0    2388      0     2388   
701   2985     3385      0    16000        1    2026      0     2026   
703   3510     3510      0    12000        1    1851      0     1851   
704   2970     3000      1    35000        1    1961      1     1481   
705   2970     3415      1        0        0    2363      0     2363   

     workscnd  exper  yngkid  yrsmarr    hrwage  agesq  
3           0     12       0       12  9.619998    900  
11          0      1       0        4  1.230000    529  
27          0     17       0        0  2.690000   1225  
42          0     21       0        0  2.159999   1936  
55          0      4       0        0  1.050000    576  
..        ...    ...     ...      ...       ...    ...  
700         0     10       0        0       NaN   1024  
701         0     27       0       18       NaN   2025  
703         0     19       0       17       NaN   1369  
704       480     31       0       22       NaN   2916  
705         0      8       1        9       NaN    900  

[306 rows x 34 columns]

In [118]:
sleep_eq1 = smf.ols(formula='sleep~totwrk+age+smsa+south', data=df).fit()
sleep_eq1.params

Intercept    3468.202663
totwrk         -0.150993
age             1.645390
smsa          -37.687393
south         136.029203
dtype: float64

In [119]:
sleep_eq1.ess.round(3)
# ESS

6300990.569

In [120]:
sleep_eq1.centered_tss.round(3)
# TSS

63490152.118

In [121]:
RSS=sleep_eq1.ssr.round(3)
RSS
# RSS

57189161.549

In [122]:
sleep_eq1.rsquared.round(7)
# R2

0.0992436

In [123]:
sleep_eq1.rsquared_adj.round(7)
# R2adj

0.0872734

In [124]:
n=306
k=4
s=np.sqrt(RSS/(n-k-1))
round(s,4)

435.8867

## 3.5

In [125]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [126]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/Labour.csv')
df

capital  labour     output       wage
0     2.606563     184   9.250759  43.080307
1     1.323237      91   3.664310  27.780016
2    22.093692     426  28.781516  44.467748
3    10.737851      72   4.124642  39.734710
4     1.161365      46   2.890150  34.650709
..         ...     ...        ...        ...
564   2.625403      20   1.424376  33.477545
565   1.276386      61   2.109048  26.300732
566   1.953869     117   6.241870  41.153979
567   1.318527      46   7.902237  66.720139
568   2.832391     161   6.804085  36.618435

[569 rows x 4 columns]

In [127]:
labour_eq1 = smf.ols(formula='output~capital+labour+wage', data=df).fit()
labour_eq1.params

Intercept   -16.745898
capital       0.395074
labour        0.049899
wage          0.436984
dtype: float64

In [128]:
labour_eq1.ess.round(3)
# ESS

2190376.077

In [129]:
labour_eq1.centered_tss.round(3)
# TSS

2235073.295

In [130]:
RSS=labour_eq1.ssr.round(3)
RSS
# RSS

44697.218

In [131]:
labour_eq1.rsquared.round(7)
# R2

0.9800019

In [132]:
labour_eq1.rsquared_adj.round(7)
# R2adj

0.9798957

In [133]:
n=569
k=3
s=np.sqrt(RSS/(n-k-1))
round(s,4)

8.8944

## 3.6

In [134]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [135]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/Labour.csv')
df

capital  labour     output       wage
0     2.606563     184   9.250759  43.080307
1     1.323237      91   3.664310  27.780016
2    22.093692     426  28.781516  44.467748
3    10.737851      72   4.124642  39.734710
4     1.161365      46   2.890150  34.650709
..         ...     ...        ...        ...
564   2.625403      20   1.424376  33.477545
565   1.276386      61   2.109048  26.300732
566   1.953869     117   6.241870  41.153979
567   1.318527      46   7.902237  66.720139
568   2.832391     161   6.804085  36.618435

[569 rows x 4 columns]

## ....

## 3.7

In [140]:
import pandas as pd
import numpy as np
import io
import statsmodels.formula.api as smf

In [141]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/Electricity.csv')
df

cost        q        pl      sl      pk      sk      pf      sf
0     0.2130      8.0   6869.47  0.3291  64.945  0.4197  18.000  0.2512
1     3.0427    869.0   8372.96  0.1030  68.227  0.2913  21.067  0.6057
2     9.4059   1412.0   7960.90  0.0891  40.692  0.1567  41.530  0.7542
3     0.7606     65.0   8971.89  0.2802  41.243  0.1282  28.539  0.5916
4     2.2587    295.0   8218.40  0.1772  71.940  0.1623  39.200  0.6606
..       ...      ...       ...     ...     ...     ...     ...     ...
153   6.8293    946.6  10642.16  0.0883  43.600  0.1914  51.463  0.7203
154   3.7605    377.0   7432.24  0.2117  74.120  0.2274  33.436  0.5609
155   3.9822    391.0   5826.04  0.1926  78.288  0.0924  44.633  0.7151
156  30.1880   5317.0   9586.63  0.0845  78.008  0.2009  41.840  0.7147
157  67.8562  15220.0   6986.09  0.1451  74.025  0.2596  25.662  0.5945

[158 rows x 8 columns]

In [142]:
elec_eq1 = smf.ols(formula='cost~q+(q**2)+pk+pf+pl', data=df).fit()
elec_eq1.params

Intercept   -84.555714
q             0.005471
pk            0.303479
pf            1.071896
pl            0.003239
dtype: float64

In [143]:
elec_eq1.ess.round(3)
# ESS

1131262.896